File: Daily_Practice/121_Best_Time_to_Buy_and_Sell_Stock.go
package main

import (
	"fmt"
)

func maxProfit(prices []int) int {
	if len(prices) == 0 {
		return 0
	}
	smallest := prices[0]
	max_profit := 0
	for _, elem := range prices {
		if elem < smallest {
			smallest = elem
		} else {
			if max_profit > elem-smallest {
				max_profit = max_profit
			} else {
				max_profit = elem - smallest
			}
		}
	}
	return max_profit
}

func main() {
	var prices = []int{5, 3, 6, 2}
	obj := maxProfit(prices)
	fmt.Println(obj)
}

File: Daily_Practice/121_Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        if(len(prices) == 0):
            return 0
        smallest = prices[0]
        max_profit = 0
        for elem in prices:
            if(elem < smallest):
                smallest = elem
            else:
                max_profit = max(max_profit, elem-smallest)
        return(max_profit)

prices = [5,3,6,2]
obj = Solution()
op = obj.maxProfit(prices)
print(op)